extensions/guacamole-vault/modules/guacamole-vault-hv/src/main/java/org/apache/guacamole/vault/hv/secret/HvSecretService.java

```diff
@@ -0,0 +1,335 @@
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.Future;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import javax.annotation.Nonnull;
+
+import org.apache.guacamole.GuacamoleException;
+import org.apache.guacamole.net.auth.Attributes;
+import org.apache.guacamole.net.auth.Connectable;
+import org.apache.guacamole.net.auth.Connection;
+import org.apache.guacamole.net.auth.ConnectionGroup;
+import org.apache.guacamole.net.auth.Directory;
+import org.apache.guacamole.net.auth.User;
+import org.apache.guacamole.net.auth.UserContext;
+import org.apache.guacamole.protocol.GuacamoleConfiguration;
+import org.apache.guacamole.token.TokenFilter;
+import org.apache.guacamole.vault.hv.GuacamoleExceptionSupplier;
+import org.apache.guacamole.vault.hv.conf.HvAttributeService;
+import org.apache.guacamole.vault.hv.conf.HvConfigurationService;
+import org.apache.guacamole.vault.hv.user.HvDirectory;
+import org.apache.guacamole.vault.secret.VaultSecretService;
+import org.apache.guacamole.vault.secret.WindowsUsername;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Service which retrieves secrets from Hashicorp Vault.
+ * The configuration used to connect to HV can be set at a global
+ * level using guacamole.properties, or using a connection group
+ * attribute.
+ */
+@Singleton
+public class HvSecretService implements VaultSecretService {
+
+    /**
+     * Logger for this class.
+     */
+    private static final Logger logger = LoggerFactory.getLogger(VaultSecretService.class);
+
+    /**
+     * Service for retrieving configuration information.
+     */
+    @Inject
+    private HvConfigurationService confService;
+
+    /**
+     * Factory for creating HV client instances.
+     */
+    @Inject
+    private HvClientFactory hvClientFactory;
+
+    /**
+     * A map of base-64 encoded JSON HV config blobs to associated HV client instances.
+     * A distinct HV client will exist for every HV config.
+     */
+    private final ConcurrentMap<String, HvClient> hvClientMap = new ConcurrentHashMap<>();
+
+    /**
+     * Create and return a HV client for the provided HV config if not already
+     * present in the client map, otherwise return the existing client entry.
+     *
+     * @param hvConfig
+     *     The base-64 encoded JSON HV config blob associated with the client entry.
+     *     If an associated entry does not already exist, it will be created using
+     *     this configuration.
+     *
+     * @return
+     *     A HV client for the provided HV config if not already present in the
+     *     client map, otherwise the existing client entry.
+     *
+     * @throws GuacamoleException
+     *     If an error occurs while creating the HV client.
+     */
+    private HvClient getClient(@Nonnull String hvConfigBase64)
```

Comment:
What is the expected behavior of the function with the `@Nonnull` annotation, if a null value is provided? This should probably be documented in the function documentation somewhere.